torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
   return std::make_shared<ProcessGroupNCCL::WorkNCCL>(devices);
 }
 
+struct checkFuture {
+  checkFuture(
+      std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+      std::vector<at::Tensor>& outputs,
+      int work_counter,
+      int input_size,
+      std::mutex& mutex)
+      : work_(std::move(work)),
```

Comment:
Just pass the future to this class since you need the future and nothing else from `work_`.